jax/interpreters/pxla.py

```diff
@@ -159,17 +159,30 @@ def replica_groups(nrep, mesh_spec, mesh_axes):
 
 ### the main pmap machinery lowers SPMD jaxprs to multi-replica XLA computations
 
-def compile_replicated(jaxpr, axis_name, axis_size, consts, *abstract_args):
-  num_replicas = axis_size * xla.jaxpr_replicas(jaxpr)
-  if num_replicas > xb.device_count():
-    msg = ("compiling computation that requires {} replicas, but only {} XLA "
-           "devices are available")
-    raise ValueError(msg.format(num_replicas, xb.device_count()))
-  axis_env = xla.AxisEnv(num_replicas, [axis_name], [axis_size])
+def compile_replicated(jaxpr, axis_name, axis_size, devices, consts,
+                       *abstract_args):
+  if devices is None:
+    num_replicas = axis_size * xla.jaxpr_replicas(jaxpr)
+    if num_replicas > xb.device_count():
+      msg = ("compiling computation that requires {} replicas, but only {} XLA "
+             "devices are available")
+      raise ValueError(msg.format(num_replicas, xb.device_count()))
+    device_assignment = None
+  else:
+    assert all(d.host_id == xb.host_id() for d in devices)
```

Comment:
Since this is checking for an error in a user-supplied argument, eventually it should probably be a ValueError rather than an assertion failure, but not blocking for this PR; just noting it for discussion and/or future ref!